corehq/apps/smsbillables/models.py

```diff
@@ -384,7 +393,7 @@ def _get_gateway_fee(cls, backend_api_id, backend_instance,
             backend_instance) or toggles.ENABLE_INCLUDE_SMS_GATEWAY_CHARGING.enabled(domain)
 
         if is_gateway_billable:
-            if backend_api_id in [SQLTwilioBackend.get_api_id(), InfobipBackend.get_api_id()]:
+            if backend_api_id in [SQLTwilioBackend.get_api_id(), InfobipBackend.get_api_id(), PinpointBackend.get_api_id()]:
```

Comment:
It seems like we're treating a handful of classes differently, and we're doing this in a couple of different places. This makes it harder to extend and easier to break.

If we refactor the special-case code into methods on the model, then we can check whether that method exists and call it when it's there, or call it and catch the AttributeError if it's missing. e.g.
```python
    try:
        status, price = backend_instance.get_provider_charges(backend_message_id)
    except AttributeError:
        raise ProviderFeeNotSupportedException(...)
```

* `utils.get_FOO_message()` -- Should this be `FooBackend.get_message()`?
* `SmsBillable._get_provider_charges` -- As above?
* `SmsBillable._get_multipart_count` -- Should this check for and call a method?
* `SmsBillable._get_gateway_fee` -- ditto

I am not going to request this as a change though, because
* I'm not super familiar with messaging on HQ, and maybe there are good reasons why this wasn't done already.
* I'm also aware that timelines are tight. I see tests are already getting pushed out. Does that mean refactors need to be pushed out too? ... If so, definitely tests and maybe a refactor should be added to a backlog somewhere.

I'll leave it to someone more familiar with messaging to confirm or contradict this suggestion.

Reply:
> call it when it's there, or call it and catch the AttributeError if it's missing. e.g. ...

Delegation is good, but I have one small disagreement here: calling a method and catching `AttributeError` may unintentionally catch the wrong error (like if the method exists, but it has a bug that causes an `AttributeError`). Better to separate the method check from the method call.
```py
if hasattr(backend_instance, "get_provider_charges"):
    status, price = backend_instance.get_provider_charges(backend_message_id)
```
Alternately, and even better, would be to implement a do-nothing method on `SQLSMSBackend` (assuming that's the base class of `backend_instance`) so you don't have to check for existence first.